storage/src/http/storage_client.rs

```diff
@@ -1258,54 +1245,38 @@ impl StorageClient {
     }
 
     async fn with_headers(&self, builder: RequestBuilder) -> Result<RequestBuilder, Error> {
-        let token = self.ts.token().await?;
+        let token = self.ts.token().await.map_err(Error::TokenSource)?;
         Ok(builder
             .header("X-Goog-Api-Client", "rust")
             .header(reqwest::header::USER_AGENT, "google-cloud-storage")
             .header(reqwest::header::AUTHORIZATION, token))
     }
 
-    async fn send<T: for<'de> serde::Deserialize<'de>>(&self, builder: RequestBuilder) -> Result<T, Error> {
+    async fn send<T>(&self, builder: RequestBuilder) -> Result<T, Error>
+    where
+        T: serde::de::DeserializeOwned,
+    {
         let request = self.with_headers(builder).await?;
         let response = request.send().await?;
-        if response.status().is_success() {
-            let full = response.bytes().await?;
-            tracing::trace!("response={:?}", &full);
-            Ok(serde_json::from_slice(&full)?)
-        } else {
-            Err(map_error(response).await)
-        }
+        let response = check_response_status(response).await?;
+        Ok(response.json().await?)
     }
 
     async fn send_get_empty(&self, builder: RequestBuilder) -> Result<(), Error> {
         let builder = self.with_headers(builder).await?;
         let response = builder.send().await?;
-        if response.status().is_success() {
-            Ok(())
-        } else {
-            Err(map_error(response).await)
-        }
+        check_response_status(response).await?;
+        Ok(())
     }
 
     async fn send_get_url(&self, builder: RequestBuilder) -> Result<String, Error> {
         let builder = self.with_headers(builder).await?;
         let response = builder.send().await?;
-        if response.status().is_success() {
-            if let Some(location) = response.headers().get(LOCATION) {
-                Ok(String::from_utf8(location.as_bytes().to_vec()).map_err(Error::FromUtf8Error)?)
-            } else {
-                Err(map_error(response).await)
-            }
-        } else {
-            Err(map_error(response).await)
-        }
+        let response = check_response_status(response).await?;
+        Ok(String::from_utf8_lossy(response.headers()[LOCATION].as_bytes()).into_owned())
```

Comment:
The use of `lossy` here is admittedly a bit of a hack, but I think it's harmless.  GCS _should_ never return non utf-8 here, and if it does it will cause a failure later when the URL is used by the resumable upload client.